components/script/dom/idbopendbrequest.rs

```diff
@@ -0,0 +1,386 @@
+            // Step 11
+            (Ok(connection), true)
+        } else {
+            // Step 11
+            (Ok(connection), false)
+        }
+    }
+
+    fn handle_delete_db(&self, result: Result<(), ()>) {
+        let open_request = self.open_request.root();
+        let global = open_request.global();
+        open_request.idbrequest.set_ready_state_done();
+
+        match result {
+            Ok(_) => {
+                let window = global.as_window();
+                let cx = window.get_cx();
+                let _ac = enter_realm(&*open_request);
+                rooted!(in(*cx) let mut answer = UndefinedValue());
+                open_request.set_result(answer.handle());
+
+                let event = Event::new(
+                    &global,
+                    Atom::from("success"),
+                    EventBubbles::DoesNotBubble,
+                    EventCancelable::NotCancelable,
+                );
+                event.upcast::<Event>().fire(open_request.upcast());
+            },
+            Err(_e) => {
+                // FIXME(rasviitanen) Set the error of request to the
+                // appropriate error
+
+                let event = Event::new(
+                    &global,
+                    Atom::from("error"),
+                    EventBubbles::Bubbles,
+                    EventCancelable::Cancelable,
+                );
+                event.upcast::<Event>().fire(open_request.upcast());
+            },
+        }
+    }
+}
+
+#[dom_struct]
+pub struct IDBOpenDBRequest {
+    idbrequest: IDBRequest,
+}
+
+impl IDBOpenDBRequest {
+    pub fn new_inherited() -> IDBOpenDBRequest {
+        IDBOpenDBRequest {
+            idbrequest: IDBRequest::new_inherited(),
+        }
+    }
+
+    pub fn new(global: &GlobalScope) -> DomRoot<IDBOpenDBRequest> {
+        reflect_dom_object(
+            Box::new(IDBOpenDBRequest::new_inherited()),
+            global,
+            IDBOpenDBRequestBinding::Wrap,
+        )
+    }
+
+    #[allow(unsafe_code)]
+    // https://www.w3.org/TR/IndexedDB-2/#run-an-upgrade-transaction
+    fn upgrade_db_version(&self, connection: &IDBDatabase, version: u64) {
+        let global = self.global();
+        // Step 2
+        let transaction = IDBTransaction::new(
+            &global.as_window(),
+            &connection,
+            IDBTransactionMode::Versionchange,
+            connection.object_stores(),
+        );
+
+        // Step 3
+        connection.set_transaction(&transaction);
+
+        // Step 4
+        transaction.set_active_flag(false);
+
+        // Step 5-7
+        let old_version = connection.version();
+        transaction.upgrade_db_version(version);
+
+        // Step 8
+        let this = Trusted::new(self);
+        let connection = Trusted::new(connection);
+        let trusted_transaction = Trusted::new(&*transaction);
+        global
+            .as_window()
+            .task_manager()
+            .dom_manipulation_task_source()
```

Comment:
According to https://www.w3.org/TR/IndexedDB-2/#database-access-task-source we should be using a new task source, and it exist for both windows and workers (so there should be a getter on GlobalScope to avoid calling as_window).